File: 2_year/MP_1C/Práctica 1(3º parte)/src/Fecha.cpp
#include "Fecha.h"
#include <iostream>
#include <string.h>

using namespace std;

Fecha::Fecha()
{
    //ctor
}

Fecha::Fecha(int dia, int mes, int agno){

    setFecha(dia, mes, agno);
}

Fecha::~Fecha()
{
    //dtor
}


void Fecha::setFecha(int dia, int mes, int agno)
{
//Checkea si el año es bisiesto o no
//-------------------------------------------------------------------------------------------
    bool bisiesto = false;
    if ((agno % 400 == 0) || (agno%4 == 0 && agno % 100 != 0))
        bisiesto = true;
//-------------------------------------------------------------------------------------------


//Corrige la fecha (Debe haber alguna manera más sencilla de hacerlo...)
//-------------------------------------------------------------------------------------------
    if(dia<1)
        dia=1;

    switch(mes)
    {
    case 1:
    case 3:
    case 5:
    case 7:
    case 8:
    case 10:
    case 12:{
                if(dia>31)
                    dia=31;
            }
            break;
    case 4:
    case 6:
    case 9:
    case 11:{
                if(dia>30)
                    dia=30;
            }
            break;
    case 2:{
                if(bisiesto==true && dia>29)
                    dia=29;
                else
                    if(bisiesto==false && dia>28)
                        dia=28;
            }
            break;
    default:{
                if(mes>12)
                        mes=12;
                else
                    if(mes<1)
                        mes=1;
            }
            break;
    }
    this->Anio=agno;
    this->Mes=mes;
    this->Dia=dia;

//-------------------------------------------------------------------------------------------

}

void Fecha::ver()const {

    if(this->Dia < 10)
        cout << "0";
    cout << this->Dia << "/";
    if(this->Mes < 10)
        cout << "0";
    cout << this->Mes << "/" << this->Anio;
}

void Fecha::ver(){

    if(this->Dia < 10)
        cout << "0";
    cout << this->Dia << "/";
    if(this->Mes < 10)
        cout << "0";
    cout << this->Mes << "/" << this->Anio;
}

bool Fecha::bisiesto() const{
    return ((this->Anio % 400 == 0) || (this->Anio%4 == 0 && this->Anio % 100 != 0));
}

Fecha Fecha::operator++(){
    *this+1;
    return *this;
}

Fecha Fecha::operator++(int flag){
    Fecha copia(*this);
    *this+1;
    return copia;
}

Fecha Fecha::operator+(int a) const
{
    Fecha cop(*this);
    bool bisiesto = false;
    if ((cop.Anio % 400 == 0) || (cop.Anio%4 == 0 && cop.Anio % 100 != 0))
        bisiesto = true;

    cop.Dia+=a;

    switch(cop.Mes)
    {
    case 1:
    case 3:
    case 5:
    case 7:
    case 8:
    case 10:
    case 12:{
                if(cop.Dia>31)
                {
                    cop.Dia-=31;
                    cop.Mes++;
                }
            }
            break;
    case 4:
    case 6:
    case 9:
    case 11:{
                if(cop.Dia>30)
                {
                    cop.Dia-=30;
                    cop.Mes++;
                }
            }
            break;
    case 2:{
                if(bisiesto==true && cop.Dia>29)
                {
                    cop.Dia-=29;
                    cop.Mes++;
                }
                else
                    if(bisiesto==false && cop.Dia>28)
                    {
                        cop.Dia-=28;
                        cop.Mes++;
                    }
            }
            break;
    }
    if(cop.Mes>12)
    {
        cop.Mes-=12;
        cop.Anio++;
    }

    return cop;
}

Fecha operator+(int a, const Fecha &F){
//Por propiedad conmutativa de la suma, le damos la vuelta y ya tenemos implementado el operador Fecha+int
    return F+a;
}

bool Fecha::operator!=(Fecha f) const
{
    bool diferentes=false;

    if(this->Anio!=f.getAnio())
        diferentes=true;
    if(this->Mes!=f.getMes())
        diferentes=true;
    if(this->Dia!=f.getDia())
        diferentes=true;

    return diferentes;
}

ostream& operator<<(ostream &o, const Fecha &f)
{
    char AbrMes[3];

    switch(f.getMes())
    {
        case 1: strcpy(AbrMes, "ene");
                break;
        case 2: strcpy(AbrMes, "feb");
                break;
        case 3: strcpy(AbrMes, "mar");
                break;
        case 4: strcpy(AbrMes, "abr");
                break;
        case 5: strcpy(AbrMes, "may");
                break;
        case 6: strcpy(AbrMes, "jun");
                break;
        case 7: strcpy(AbrMes, "jul");
                break;
        case 8: strcpy(AbrMes, "ago");
                break;
        case 9: strcpy(AbrMes, "sep");
                break;
        case 10:strcpy(AbrMes, "oct");
                break;
        case 11:strcpy(AbrMes, "nov");
                break;
        case 12:strcpy(AbrMes, "dic");
                break;
    }
    o << (f.getDia()<10?"0":"") << f.getDia() << " " << AbrMes << " " << f.getAnio();

    return o;
}

File: 2_year/MP_1C/Práctica 2/Práctica2_Java/src/libClases/Fecha.java
package libClases;
import java.util.Scanner;

public final class Fecha implements Cloneable, Proceso {
	//ATRIBUTOS
	private int Dia, Mes, Anio;
	
	//MÉTODOS
	public Fecha(int d, int m, int a)
	{
		setFecha(d, m, a);
	}
	public Fecha(Fecha f)
	{
		this(f.Dia, f.Mes, f.Anio);
	}
	
	public int getDia()
	{
		return Dia;
	}
	public int getMes()
	{
		return Mes;
	}
	public int getAnio() 
	{
		return Anio;
	}
	public void setFecha(int d, int m, int a)//Clase mutable
	{
		//Checkea si el año es bisiesto o no
		//--------------------------------------------------------------------------------
		    boolean bisiesto = false;
		    if ((a % 400 == 0) || (a%4 == 0 && a % 100 != 0))
		        bisiesto = true;
		//---------------------------------------------------------------------------------


		//Corrige la fecha (Debe haber alguna manera más sencilla de hacerlo...)
		//---------------------------------------------------------------------------------
		    if(d<1)
		        d=1;

		    switch(m)
		    {
		    case 1:
		    case 3:
		    case 5:
		    case 7:
		    case 8:
		    case 10:
		    case 12:{
		                if(d>31)
		                    d=31;
		            }
		            break;
		    case 4:
		    case 6:
		    case 9:
		    case 11:{
		                if(d>30)
		                    d=30;
		            }
		            break;
		    case 2:{
		                if(bisiesto==true && d>29)
		                    d=29;
		                else
		                    if(bisiesto==false && d>28)
		                        d=28;
		            }
		            break;
		    default:{
		                if(m>12)
		                        m=12;
		                else
		                    if(m<1)
		                        m=1;
		            }
		            break;
		    }
		    this.Anio=a;
		    this.Mes=m;
		    this.Dia=d;

		//---------------------------------------------------------------------------------
	}
	public void setDia(int dia)
	{
		this.Dia=dia;
	}
	public void setMes(int mes)
	{
		this.Mes=mes;
	}
	public void setAnio(int anio)
	{
		this.Anio=anio;
	}
	
	public static Fecha pedirFecha() 
	{
		//SI SE INTRODUCE FECHA NO VALIDA, SE PIDE DE NUEVO HASTA QUE SE INTRODUZCA CORRECTAMENTE
		int d, m, a;
		Scanner Lector = new Scanner (System.in);
		boolean fechaValida;
		
		do
		{
			//Creamos una string que contenga los datos y los divida en 3 sub-arrays divididos
			//por el caracter "/", cada uno contendrá un dato de la fecha.
			System.out.println("Introduce Fecha (dd/mm/yyyy): ");
			String[] fechaIntroducida = Lector.nextLine().split("/");
			d = Integer.parseInt(fechaIntroducida[0]);
			m = Integer.parseInt(fechaIntroducida[1]);
			a = Integer.parseInt(fechaIntroducida[2]);
			
			//COMPROBAMOS QUE LA FECHA INTRODUCIDA POR TECLADO ES VALIDA
			//Es el mismo código usado para corregir pero un poco alterado
			//----------------------------------------------------------
			
			//Checkea si el año es bisiesto o no
			//--------------------------------------------------------------------------------
			    boolean bisiesto = false;
			    if ((a % 400 == 0) || (a%4 == 0 && a % 100 != 0))
			        bisiesto = true;
			//---------------------------------------------------------------------------------


			//Comprueba que la fecha es valida (Se puede simplificar...)
			//---------------------------------------------------------------------------------
			    fechaValida = true;
			    
			    if(d<1)
			        fechaValida=false;

			    switch(m)
			    {
			    case 1:
			    case 3:
			    case 5:
			    case 7:
			    case 8:
			    case 10:
			    case 12:{
			                if(d>31)
			                    fechaValida=false;
			            }
			            break;
			    case 4:
			    case 6:
			    case 9:
			    case 11:{
			                if(d>30)
			                    fechaValida=false;
			            }
			            break;
			    case 2:{
			                if(bisiesto==true && d>29)
			                    fechaValida=false;
			                else if(bisiesto==false && d>28)
			                        fechaValida=false;
			            }
			            break;
			    default:{
			                if(m>12)
			                        fechaValida=false;
			                else
			                    if(m<1)
			                        fechaValida=false;
			            }
			            break;
			    }
			//---------------------------------------------------------------------------------
			if(fechaValida==false)
				System.out.println("Fecha no valida, introduzca de nuevo\n");
			
		}while(fechaValida==false);
		
		//Una vez la fecha es introducida correctamente, se construye su instancia y se devuelve.
		Fecha f = new Fecha(d,m,a);
		return f;
	}
	public static boolean mayor(Fecha f2, Fecha f1)
	{
		boolean mayorPrimero=false;
		
		if(f2.Anio>f1.Anio)
			mayorPrimero=true;
		if(f2.Anio == f1.Anio && f2.Mes > f1.Mes)
			mayorPrimero=true;
		if(f2.Anio == f1.Anio && f2.Mes == f1.Mes && f2.Dia > f1.Dia)
			mayorPrimero=true;
		
		return mayorPrimero;
	}
	
	@Override
	public String toString()
	{
		return((Dia<10? "0":"")+Dia+"/"+(Mes<10? "0":"")+Mes+"/"+Anio);
	}
	
	@Override
	public Object clone()
	{
		//Código para copias binarias de objetos
		Object obj=null;
		try
		{
			obj=super.clone();
		}catch(CloneNotSupportedException ex)
		{
			System.out.println("No se puede duplicar");
		}
		
		return obj;
	}
	
	@Override
	public boolean equals(Object obj)
	{
		boolean iguales=false;
		Fecha f;
		
		if((obj instanceof Fecha))
		{
			f = (Fecha)obj;
			if(Dia==f.Dia && Mes==f.Mes && Anio==f.Anio)
				iguales = true;
		}
		
		return iguales;
	}
	
	@Override
	public void ver()
	{
		System.out.println(this);
	}
	
	public boolean bisiesto()
	{
	    return ((Anio % 400 == 0) || (Anio%4 == 0 && Anio % 100 != 0));
	    //El operador == compara las direcciones de memoria ¡CUIDADO!
	    //Pero para tipos primitivos NO
	}
	
	public Fecha diaSig()
	{
		Fecha f = (Fecha)clone();
		
		f.Dia++;
		
		//Comprueba que no se ha sobrepasado el número de días del mes respectivo
		//-----------------------------------------------------------------------
		switch(f.Mes)
		{
		case 1:
		case 3:
		case 5:
		case 7:
		case 8:
		case 10:
		case 12: 
		{
			if(f.Dia==32)
		    {
				f.Dia-=31;
				f.Mes++;
	   	    }
		}break;
		case 4:
		case 6:
		case 9:
		case 11:
		{
			if(f.Dia==31)
		    {
				f.Dia-=30;
				f.Mes++;
	   	    }
		}break;
		case 2:
		{
			if(f.Dia==30 && f.bisiesto())
		    {
				f.Dia-=29;
				f.Mes++;
	   	    }
			else if(f.Dia==29 && !f.bisiesto())
			{
				f.Dia-=28;
				f.Mes++;
			}
		}
		}
		
		if(f.Mes==13)
		{
			f.Mes-=12;
			f.Anio++;
		}
		//-----------------------------------------------------------------------
		
		return f;
	}

	public static void main(String[] args) 
	{
		Fecha f1 = new Fecha(29,2,2001), f2 = new Fecha(f1), f3 = new Fecha(29,2,2004);
		final Fecha f4=new Fecha(05,12,2023); //Es constante la referencia f4
		
		System.out.println("Fechas: " + f1.toString() + ", "+f2+ ", " +f3+ ", " +f4+ "\n");
		
		f1=new Fecha(31,12,2016); //31/12/2016
		f4.setFecha(28, 2, 2008); //Pero no es constante el objeto al que apunta
		
		System.out.println(f1 +" "+ f2.toString() +" " + f3 +" "+ f4+" "+ f1);
		
		f1=new Fecha(f4.getDia()-10, f4.getMes(), f4.getAnio()-7); //f1=18/02/2001
		f3=Fecha.pedirFecha(); //Pide una fecha por teclado
		
		if (f3.bisiesto() && Fecha.mayor(f2,f1))
			System.out.print("El " + f3.getAnio() + " fue bisiesto, " + f1 + ", " + f3);

	}
}
